// rah.auth/index.js
var jwt = require('express-jwt');
var settings = require('rah.utils').settings;
var crypto = require('rah.utils').crypto;

function createToken(data) {
    var token = jwt.sign({ data }, settings.token.secret);
    return token;
}

function validToken(type, module) {
    return jwt(
        {
            secret: settings.token.secret,
            getToken: function (req) {
                var token = null;
                switch (type) {
                    case 'header':
                        token = req.headers['x-access-token'];
                    case 'cookie':
                        token = req.cookies['x-access-token'];
                }

                return token;
            }
        });
}

function validCrypto(req, res, next) {
    console.log('teste');
    res.header("Access-Control-Allow-Origin", "https://seugarcom.herokuapp.com");
    res.header("Access-Control-Allow-Headers", "Origin, X-Requested-With, Content-Type, Accept");
    res.header("Access-Control-Allow-Methods", "POST, GET, PUT, DELETE, OPTIONS");
    res.header("Access-Control-Allow-Credentials", "true");
    console.log('teste3');


    if (req.body.crypto) {
        req.body = crypto.decrypt(req.body.crypto);
    }
    next();
}

module.exports = {
    createToken: function (data) { return createToken(data); },
    req: validToken('header'),
    cookie: validToken('cookie'),
    crypto: validCrypto
}
